horovod/common/operations.cc

```diff
@@ -1467,14 +1482,38 @@ void BackgroundThreadLoop(HorovodGlobalState& state) {
     (*it)(SHUT_DOWN_ERROR);
   }
 
-  MPI_Finalize();
+  // Calls MPI_Finalize if horovod_global.auto_mpi_finalize is set otherwise prints warning
+  while(true){
```

Comment:
I'm not sure this actually works.  What needs to happen is `horovod_finalize()` should have similar logic to `~HorovodGlobalState()` - we need to set `shut_down` flag and wait for the background thread to terminate.